modules/core/src/main/java/org/apache/ignite/internal/processors/cache/distributed/dht/colocated/SavepointUnlockFuture.java

```diff
@@ -0,0 +1,487 @@
+import org.apache.ignite.internal.processors.cache.version.GridCacheVersion;
+import org.apache.ignite.internal.util.future.GridFutureAdapter;
+import org.apache.ignite.internal.util.tostring.GridToStringExclude;
+import org.apache.ignite.internal.util.tostring.GridToStringInclude;
+import org.apache.ignite.internal.util.typedef.C1;
+import org.apache.ignite.internal.util.typedef.F;
+import org.apache.ignite.internal.util.typedef.internal.CU;
+import org.apache.ignite.internal.util.typedef.internal.S;
+import org.apache.ignite.internal.util.typedef.internal.U;
+import org.apache.ignite.lang.IgniteUuid;
+
+/**
+ * Colocated cache unlock during savepoint future.
+ */
+public final class SavepointUnlockFuture extends GridCacheCompoundIdentityFuture<Boolean> {
+    /** */
+    private static final long serialVersionUID = 0L;
+
+    /** Logger reference. */
+    private static final AtomicReference<IgniteLogger> logRef = new AtomicReference<>();
+
+    /** Logger. */
+    private static IgniteLogger log;
+
+    /** Logger. */
+    private static IgniteLogger msgLog;
+
+    /** Cache registry. */
+    @GridToStringExclude
+    private final GridCacheContext<?, ?> ctx;
+
+    /** Future ID. */
+    private final IgniteUuid futId;
+
+    /** Lock version. */
+    private final GridCacheVersion lockVer;
+
+    /** Error. */
+    private volatile Throwable err;
+
+    /** Transaction. */
+    @GridToStringExclude
+    private final GridNearTxLocal tx;
+
+    /** Key to unlock */
+    private final Collection<IgniteTxKey> keys;
+
+    /** Trackable flag (here may be non-volatile). */
+    private boolean trackable;
+
+    /**
+     * @param ctx Registry.
+     * @param tx Transaction.
+     */
+    public SavepointUnlockFuture(GridCacheContext<?, ?> ctx, GridNearTxLocal tx) {
+        super(CU.boolReducer());
+
+        this.ctx = ctx;
+        this.tx = tx;
+        this.keys = new ArrayList<>();
+
+        ignoreInterrupts();
+
+        lockVer = tx.xidVersion();
+
+        futId = IgniteUuid.randomUuid();
+
+        if (log == null) {
+            msgLog = ctx.shared().txLockMessageLogger();
+            log = U.logger(ctx.kernalContext(), logRef, SavepointUnlockFuture.class);
+        }
+    }
+
+    /**
+     * @return Future ID.
+     */
+    @Override public IgniteUuid futureId() {
+        return futId;
+    }
+
+    /** {@inheritDoc} */
+    @Override public boolean trackable() {
+        return trackable;
+    }
+
+    /** {@inheritDoc} */
+    @Override public void markNotTrackable() {
+        trackable = false;
+    }
+
+    /** {@inheritDoc} */
+    @Override public boolean onNodeLeft(UUID nodeId) {
+        boolean found = false;
+
+        for (IgniteInternalFuture<?> fut : futures()) {
+            if (isMini(fut)) {
+                MiniFuture f = (MiniFuture)fut;
+
+                if (f.node().id().equals(nodeId)) {
+                    f.onResult(newTopologyException(nodeId));
+
+                    found = true;
```

Comment:
Can be several MiniFuture for one node? Or we can `break` here?

Reply:
Can't, we can break here.